### ISSUE TITLE
Upstream scale factor issues

### ISSUE THREAD

**Issue by JasonLG1979** [pithos/pithos#552]
Currently adjusting the scale factor(in GNOME Shell at least) breaks our Gtk.CellRenderer. To add to that Gtk.Widget.get_scale_factor is broken and always returns 1 so it's useless. Eventually I plan on redesigning the UI and getting rid of the Gtk.CellRenderer but that doesn't fix pixbuf scaling. With an off the shelf Gtk.Image pixbufs still don't scale in response to a scale factor change, SVG icons do however.

I think I have a workaround:
https://gist.github.com/JasonLG1979/c23206f80b1d7750e84f3f8912d7c812

Basically you would pass the uri of an image to SmartImage.set_pixbuf_from_uri and it would cache the original pixbuf and scale itself and the pixbuf when needed. it also falls back to a fallback icon of your choice instead of a broken image icon. I think it could be very useful for use as the cover images in the new UI design.

@TingPing check it out and tell me what you think.

**Comment by JasonLG1979** [comment]
Oh and SmartImage assumes a square Gtk.Image area and keeps the original image's aspect ratio within that square area.

**Comment by JasonLG1979** [comment]
You can test it by adjusting the scale factor while the window is open. The image should scale to the scale factor.

**Comment by TingPing** [comment]
> Currently adjusting the scale factor(in GNOME Shell at least)

Where exactly are you setting it?

*Gtk.Image* is a higher level API and should handle all scaling for you.

**Comment by JasonLG1979** [comment]
Scratch that Gtk.Image doesn't handle scale factor at all. Adjusting the scale factor changes all UI element sizes except icons and images... Geez scale factor support in Gtk/GNOME Shell really sucks apparently...

Anyway I've updated that Gist. Using my SmartImage icons and images scale properly with the rest of the UI elements when you change the scale factor.

Here's some screenshots. The smart images are the ones on the ends.

At 1.0 everything looks normal ofc (except that the default scaling in the pixbuf in the Gtk.Image sucks):
![screenshot from 2017-12-04 11-07-47](https://user-images.githubusercontent.com/6667703/33565776-a92f62c4-d8e3-11e7-8dc0-1d92bd2ee2c8.png)

And at 2.0 you can see that other elements are scaled but not the default Gtk.Images:
![screenshot from 2017-12-04 11-08-11](https://user-images.githubusercontent.com/6667703/33565816-c59dd724-d8e3-11e7-8cf2-b9bc0ee1ced2.png)

**Comment by JasonLG1979** [comment]
As a reference, here's what Pithos looks like at 2.0:
![screenshot from 2017-12-04 11-11-55](https://user-images.githubusercontent.com/6667703/33565940-26aa8a62-d8e4-11e7-96b3-f26a0066d5ff.png)

Not pretty...

**Comment by JasonLG1979** [comment]
I had similar issues with scaling with my Shell Extension. I guess I really don't understand why it's so hard for upstream to fix this. Design your widgets assuming a fixed size as a baseline for scaling purposes (say 96px = 1 inch on screen for example) and then get the pixel density of the screen,  and then divide them to get a scale factor and then multiply that again by your baseline widget size to get  it's actual size. They already do that kinda except they don't implement the actual scaling.

For example(totally made up numbers):
baseline = 96
pixel_density = 124
unscaled_widget_size = 63
actual_widget_size =  round((pixel_density /  baseline) * unscaled_widget_size)

You could either have a fixed assumed baseline or let the theme creators set the baseline. So they could basically say "hey this theme looks best at this pixel_density, use it as the baseline for scaling purposes."

**Comment by TingPing** [comment]
> It handles GIcons and icon names just fine but doesn't scale pixbufs. I'd guess it's a pixbuf issue not a Gtk.Image issue. Anyway you wouldn't want to re-scale an already scaled raster image. It would look like crap. You need a cached copy of the pixbuf so that when you need to rescale you rescale the original.

Indeed.

> In GNOME Tweak Tool.

I do believe that is considered a bit of a hack. GNOME-Shell has its own internal mechanisms for determining scale since it varies by display. I'm not sure the details though. Try the `gdk_window_get_scale_factor()` API?

**Comment by JasonLG1979** [comment]
All get_scale_factor methods in Gtk return 1 always no matter what.

**Comment by TingPing** [comment]
In GNOME Tweak 3.27 I don't even see any scaling option beyond fonts. Using the scaling option inside the Gtk Inspector works correctly.

**Comment by TingPing** [comment]
`env GDK_SCALE=2 pithos` also works perfectly here.

**Comment by TingPing** [comment]
Ah looking at your screenshot that is indeed the confusion, you are scaling fonts but nothing else, so the image is small because the text next to it is just so tall.

**Comment by JasonLG1979** [comment]
Scaling fonts scales every element that contains fonts which is pretty much all elements except elements that contain images.

**Comment by JasonLG1979** [comment]
In any event things are defiantly broken.

**Comment by JasonLG1979** [comment]
That was my problem with my extension also. Inconsistent scaling.

**Comment by TingPing** [comment]
I see no problem, you scaled only fonts and then complained only fonts scaled.
